[ROVER-362] Preserve cost inflation filter

ErikCald opened the pull request:
Created an grid map filter that inflates a layer while preserving the cost. It will inflate each cell to the surrounding cells in a circle, always picking the worse traversability value. The implementation of this filter is not very efficent. It works with my 40x40 gridmap, only taking 0.001 seconds on my jetson orin nano. If we want to scale that up I will need to make it more efficient or use multi processing.

### Package cprt_gridmap_filters
* Added a new package for gridmap filters to make it easier to create a specific CMakeLists.txt. Can probably merge it into another package if we want to consolidate pacakges.
* Added PreserveCostInflationFilter to this package. Uses the gridmap library and eigen library.

### Configuration and Launch Files
* Created config files in the navigation package for elevation_mapping and the traversability pipeline. Put into navigation since this doesn't really belong in the elevation_mapping repo itself as it will be specific to our rover.
* Created launch files in the navigation package to make it easy to launch the traversability gridmap and the zed camera.
  * Launch traversability gridmap with: `ros2 launch navigation traversability_gridmap.launch.py`
  * Launch zed node with: `ros2 launch navigation zed.launch.py`
  * Or Launch zed node with svo path with: `ros2 launch navigation zed.launch.py svo_path:=~/Documents/ZED/WoodBoardSlopeTest_Dec4/HD720_SN32985835_19-39-17.svo2`

### Rviz config 
* Created an rviz config for the traversibility gridmap. 
  * `cd cprt_rover_24/src/navigation/rviz`
  * `rviz2 -d traversibility_gridmap.rviz`


### Images
Here are some images of what it looks like. 
* The traversability layer is shown on the lower layer. The height of this layer is the elevation and the colour represents traverability as a number from 0 to 1 where 1 is good and 0 is bad.
* The inflated traversability layer is a flat layer above the traversability. It has the same 0 to 1 number. 
* The settings for these images are: 
  * HD720 60fps, point clouds are published at 10fps and downsampled by 3. Zed depth is limited to 9 meters.
  * Center point of the map is 5 meters forward and 0.5 meters down from the zed camera
  * Points 0.5 meters above the zed camera are ignored
  * Gridmap size is 10m x 10m
  * Gridmap resolution is 0.25m (the lines on each layer show 0.25m)
  * The inflated layer is inflated radial by 0.5m
![WoodSlopeTest-TraversibilityInflated-Jan20-Img3](https://github.com/user-attachments/assets/cc8ae3d9-80c7-40a7-914b-fdd152cf3964)
![WoodSlopeTest-TraversibilityInflated-Jan20-Img](https://github.com/user-attachments/assets/256be7de-2de7-42a4-a633-5b18d322c4f0)
![WoodSlopeTest-TraversibilityInflated-Jan20-Img1](https://github.com/user-attachments/assets/28049229-4c1d-4405-8ef2-c88a835c2cee)

ConnorNeed commented on src/cprt_gridmap_filters/src/PreserveCostInflationFilter.cpp:
```diff
@@ -0,0 +1,129 @@
+  for (grid_map::CircleIterator iterator(mapOut, position,
+                                         this->inflationRadius_);
+       !iterator.isPastEnd(); ++iterator) {
+    if (!mapOut.isValid(*iterator, this->outputLayer_)) {
+      continue;
+    }
+
+    layerOut.coeffRef((*iterator)(0), (*iterator)(1)) =
+        std::min(layerOut.coeff((*iterator)(0), (*iterator)(1)), value);
```
Does this work? It might be slightly more effiecient. Will need to include cmath
```suggestion
    const std::size_t index = iterator.getLinearIndex();
    auto & cellValue = layerOut(i);
    if (!std::isfinite(cellValue)){
      continue;
    }
    cellValue = std::min(cellValue, value);
```

ErikCald replied:
> Does this work? It might be slightly more effiecient. Will need to include cmath

The circle iterator doesn't have a getLinearIndex, and the layerOut(i) returns a Eigen::IndexedView which isn't compiling with the next function calls.

But I've updated it to use the reference to the cell value using the at function which is what they show in this demo file: https://github.com/ANYbotics/grid_map/blob/be3bbc9afbcebd6a5cadcb091211916a79d3dc14/grid_map_demos/src/IteratorsDemo.cpp#L91

ConnorNeed commented on src/cprt_gridmap_filters/src/PreserveCostInflationFilter.cpp:
```diff
@@ -0,0 +1,129 @@
+  for (grid_map::CircleIterator iterator(mapOut, position,
+                                         this->inflationRadius_);
+       !iterator.isPastEnd(); ++iterator) {
+    if (!mapOut.isValid(*iterator, this->outputLayer_)) {
+      continue;
+    }
+
+    layerOut.coeffRef((*iterator)(0), (*iterator)(1)) =
+        std::min(layerOut.coeff((*iterator)(0), (*iterator)(1)), value);
```
Why is it min and not max?

ErikCald replied:
It's currently min because the traversability is set to `0 == not traversable` and `1 == traversable`. But I'm thinking of swapping them around because it makes more sense to have 0 as not blocked and 1 as blocked, and it makes the inflation filter more adaptable to other projects too.
I'll try this, test on my jetson and reply back here.

ConnorNeed commented on src/cprt_gridmap_filters/src/ThresholdFilter.cpp:
What did you change with this filter?

ErikCald replied:
I didn't change anything in this filter, and I don't currently use it in any code. I copied it from the gridmap repo to test the CMakeLists.txt and other build files are functioning correctly. I can remove it, or leave it as a `unit test` for future code/build changes.

ConnorNeed replied:
I think we can probably remove it since we got this new one building.

ErikCald replied:
Removed it

ConnorNeed commented on src/navigation/config/elevation_mapping/zed2i_sensor_processor.yaml:
```diff
@@ -0,0 +1,13 @@
+    type: structured_light
+    cutoff_min_depth: 0.1
+    cutoff_max_depth: 35.0
+    normal_factor_a: 0.000611
```
Are these Distortion coefficients?

ErikCald replied:
No not quite. I can't find the resource I found the last time that I looked into these settings. The outcome from that research was these values are difficult to calibrate/measure and that the values I have here copied from the RealSense D435 are more than good enough for the ZED camera. 

The ZED SDK already rectifies the image and gives a point cloud, the elevation_mapping only takes a pointcloud so I don't really know what it does with these values.

ConnorNeed reviewed (commented):
Really good stuff! What's the top value when running on your jetson Nano?

ErikCald commented:
> Really good stuff! What's the top value when running on your jetson Nano?

The elevation mapping is taking less than 10% of my cpu with this 40x40 gridmap.
![ErikJetsonOrinNanoJan24-TopOfElevationMapping](https://github.com/user-attachments/assets/f66d3663-58dd-4a06-a8a5-32d659adc55e)


I changed the resolution from 0.25m to 0.1m and changed the zed point cloud down sampling from 3 to 2 (720p downsampled by 2). That means the 10m x 10m grid map is 100 x 100. When the gridmap has this finer resolution, the point cloud is not dense enough and there are alot of invalid points farther away as you can see in this image:
![HigherResGridmapAndHalfSizePointCloud-Jan24-Img1](https://github.com/user-attachments/assets/47c46566-9108-40ff-942b-69799edb7e37)

The top for this 100 x 100 gridmap: The elevation mapping was between 10-50 %CPU. It's weird that it has such a wide range.

I tried changing the zed downsampling to 1. Here is what that looks like with the 100 x 100 gridmap still:
![Jan24-GridmapImage-NoDownsampling](https://github.com/user-attachments/assets/8bb0af78-5206-4b6b-bfd3-940575ab5684)

ConnorNeed reviewed (approved):
LGTM
